astTraversal/tags.go

```diff
@@ -5,46 +5,88 @@ import (
 	"strings"
 )
 
-func ParseStructTag(tag string) (name string, isRequired bool, isShown bool) {
-	isShown = true
-	if tag != "" {
-		binding := reflect.StructTag(tag).Get("binding")
-		if binding != "" {
-			isRequired = strings.Contains(binding, "required")
+type BindingTagType string
+
+const (
+	HeaderBindingTag BindingTagType = "header"
+	FormBindingTag   BindingTagType = "form"
+	URIBindingTag    BindingTagType = "uri"
+	JSONBindingTag   BindingTagType = "json"
+	XMLBindingTag    BindingTagType = "xml"
+	YAMLBindingTag   BindingTagType = "yaml"
+	NoBindingTag     BindingTagType = ""
+)
+
+var BindingTags = []BindingTagType{HeaderBindingTag, FormBindingTag, URIBindingTag, JSONBindingTag, XMLBindingTag, YAMLBindingTag}
+
+type BindingTag struct {
+	Name       string
+	IsShown    bool
+	IsOptional bool
+}
+
+type BindingTagMap map[BindingTagType]BindingTag
+
+type ValidationTagType string
+
+const (
+	GinValidationTag       ValidationTagType = "binding"
+	ValidatorValidationTag ValidationTagType = "validate"
+	NoValidationTag        ValidationTagType = ""
+)
+
+var ValidationTags = []ValidationTagType{GinValidationTag, ValidatorValidationTag}
+
+type ValidationTag struct {
+	IsRequired bool
+}
+
+type ValidationTagMap map[ValidationTagType]ValidationTag
+
+func ParseStructTag(field string, tag string) (BindingTagMap, ValidationTagMap) {
+	bindingTags := make(BindingTagMap)
+	for _, bindingTag := range BindingTags {
+		newBindingTag := BindingTag{}
+
+		tagValue := reflect.StructTag(tag).Get(string(bindingTag))
+		if tagValue == "" {
+			continue
 		}
 
-		yaml := reflect.StructTag(tag).Get("yaml")
-		if yaml != "" && yaml != "-" {
-			isShown = true
-			name = strings.Split(yaml, ",")[0]
-		} else if yaml == "-" && isShown {
-			isShown = false
+		tagItems := strings.Split(tagValue, ",")
+		if tagItems[0] == "" {
+			newBindingTag.Name = field
+			newBindingTag.IsShown = true
+		} else if tagItems[0] != "-" {
+			newBindingTag.Name = tagItems[0]
+			newBindingTag.IsShown = true
+		} else {
+			newBindingTag.IsShown = false
 		}
 
-		xml := reflect.StructTag(tag).Get("xml")
-		if xml != "" && xml != "-" {
-			isShown = true
-			name = strings.Split(xml, ",")[0]
-		} else if xml == "-" && isShown {
-			isShown = false
+		if len(tagItems) > 1 && tagItems[1] == "omitempty" {
```

Comment:
Is this how we want do optionality? Not binding:required?

Reply:
Well I think here it's a question of what we mean by optionality because everything is optionally unless there is a binding:required so I get what you mean but here what I meant by optional was if it is optional in the return type of the struct i.e. omitempty

Reply:
Yeah I see the logic, like we make it required by default implicitly, then if they want it to be optional we are specific about adding the omit empty. Can you add this to docs?
Also what does binding required do in our flow then?

Reply:
Yeah I will do. That's a good point actually I don't think it does much at the moment.